File: back/app/schemas/evals.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class EvalCaseCreate(BaseModel):
    query: str = Field(..., min_length=1, description="评测问题")
    expected_doc: str | None = Field(None, description="期望命中文档路径或标题")
    expected_points: str | None = Field(None, description="期望要点 / 关键词")
    enabled: bool = True


class EvalCaseUpdate(BaseModel):
    query: str | None = Field(None, min_length=1)
    expected_doc: str | None = None
    expected_points: str | None = None
    enabled: bool | None = None


class EvalCaseData(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    query: str
    expected_doc: str | None
    expected_points: str | None
    enabled: bool
    created_at: datetime
    updated_at: datetime


class EvalCaseListData(BaseModel):
    total: int
    items: list[EvalCaseData]


class EvalRunCreate(BaseModel):
    name: str = Field("", description="测评任务名称")
    remark: str | None = None
    case_ids: list[int] = Field(..., min_length=1, description="勾选的题库用例 id")


class EvalRunItemData(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    run_id: int
    case_id: int | None
    query_snapshot: str
    expected_doc_snapshot: str | None
    expected_points_snapshot: str | None
    optimized_query: str | None
    hits_json: str | None
    answer: str | None
    ask_status: str | None
    duration_ms: int | None
    rewrite_fallback: bool
    auto_score: int | None
    auto_reason: str | None
    needs_review: bool
    human_score: int | None
    final_score: int | None
    human_comment: str | None
    passed: bool | None


class EvalRunData(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    remark: str | None
    status: str
    model: str | None
    total: int
    passed: int
    needs_review: int
    error_count: int
    avg_score: float | None
    created_by: int | None
    created_at: datetime
    finished_at: datetime | None


class EvalRunDetailData(EvalRunData):
    items: list[EvalRunItemData] = []


class EvalRunListData(BaseModel):
    total: int
    items: list[EvalRunData]


class EvalHumanScoreRequest(BaseModel):
    human_score: int = Field(..., ge=0, le=100)
    human_comment: str | None = None

File: front/src/api/evals.ts
import { http } from './request'

export type EvalCase = {
  id: number
  query: string
  expected_doc: string | null
  expected_points: string | null
  enabled: boolean
  created_at: string
  updated_at: string
}

export type EvalCaseList = {
  total: number
  items: EvalCase[]
}

export type EvalRunItem = {
  id: number
  run_id: number
  case_id: number | null
  query_snapshot: string
  expected_doc_snapshot: string | null
  expected_points_snapshot: string | null
  optimized_query: string | null
  hits_json: string | null
  answer: string | null
  ask_status: string | null
  duration_ms: number | null
  rewrite_fallback: boolean
  auto_score: number | null
  auto_reason: string | null
  needs_review: boolean
  human_score: number | null
  final_score: number | null
  human_comment: string | null
  passed: boolean | null
}

export type EvalRun = {
  id: number
  name: string
  remark: string | null
  status: 'pending' | 'running' | 'done' | string
  model: string | null
  total: number
  passed: number
  needs_review: number
  error_count: number
  avg_score: number | null
  created_by: number | null
  created_at: string
  finished_at: string | null
}

export type EvalRunDetail = EvalRun & {
  items: EvalRunItem[]
}

export type EvalRunList = {
  total: number
  items: EvalRun[]
}

export const EVAL_STATUS_LABEL: Record<string, string> = {
  pending: '待测试',
  running: '测评中',
  done: '已测试',
}

export function listEvalCases(params?: {
  q?: string
  enabled?: boolean
  offset?: number
  limit?: number
}) {
  return http.get<EvalCaseList>('/evals/cases', { params })
}

export function createEvalCase(data: {
  query: string
  expected_doc?: string
  expected_points?: string
  enabled?: boolean
}) {
  return http.post<EvalCase>('/evals/cases', data)
}

export function updateEvalCase(
  id: number,
  data: Partial<{
    query: string
    expected_doc: string | null
    expected_points: string | null
    enabled: boolean
  }>,
) {
  return http.patch<EvalCase>(`/evals/cases/${id}`, data)
}

export function deleteEvalCase(id: number) {
  return http.delete<null>(`/evals/cases/${id}`)
}

export function listEvalRuns(params?: { offset?: number; limit?: number }) {
  return http.get<EvalRunList>('/evals/runs', { params })
}

export function createEvalRun(data: {
  name?: string
  remark?: string
  case_ids: number[]
}) {
  return http.post<EvalRunDetail>('/evals/runs', data)
}

export function getEvalRun(id: number) {
  return http.get<EvalRunDetail>(`/evals/runs/${id}`)
}

export function startEvalRun(id: number) {
  // 逐题调用 Ask，可能较久
  return http.post<EvalRunDetail>(`/evals/runs/${id}/start`, undefined, {
    timeout: 600_000,
  })
}

export function deleteEvalRun(id: number) {
  return http.delete<null>(`/evals/runs/${id}`)
}

export function scoreEvalRunItem(
  runId: number,
  itemId: number,
  data: { human_score: number; human_comment?: string },
) {
  return http.patch<EvalRunItem>(`/evals/runs/${runId}/items/${itemId}/score`, data)
}
